File: src/main/java/com/example/eventmanager/service/EventService.java
package com.example.eventmanager.service;


import com.example.eventmanager.dao.EventRepository;
import com.example.eventmanager.domain.CalendarData;
import com.example.eventmanager.domain.Category;
import com.example.eventmanager.domain.Event;
import com.example.eventmanager.domain.User;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;


import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Service
public class EventService {

    private final EventRepository eventRepository;
    private final UserService userService;

    @Autowired
    public EventService(EventRepository eventRepository, UserService userService) {
        this.eventRepository = eventRepository;
        this.userService = userService;
    }


    public void createEvent(Event event){
        event.setId((long) eventRepository.save(event));
        eventRepository.addUserToEvent(userService.getCurrentUser().getId(),event.getId());
    }


    public void updateEvent(Event event){
        eventRepository.update(event);
    }

    public Event getEvent (Long id){
        return eventRepository.findOne(id);
    }

    public List<Event> getUserEvents(){

       return eventRepository.findAllUserEvents(userService.getCurrentUser().getId());
    }

    public List<Event> getEventsWithUserParticipation(Long userId,Boolean isPrivate,Boolean isSent){

        return eventRepository.findEventsWithUserParticipation(userId, isPrivate, isSent);
    }

    public List<Event> getAllPublicEvents(){

        return eventRepository.findAllPublicEvents();
    }

    public List<Event> eventsForPeriod(Long user_id,LocalDate fromDate, LocalDate toDate){

        return eventRepository.eventsForPeriod(user_id,fromDate,toDate);
    }

    public void joinToEvent(Long event_id){

        eventRepository.addUserToEvent(userService.getCurrentUser().getId(),event_id);

    }

    public void AddUsersToEvent(List<User> users,Long id){

        for (User user:users) {
            eventRepository.addUserToEvent(user.getId(),id);
        }
    }

    public List<User> getParticipants(Long id){

        return eventRepository.findParticipants(id);
    }

    public void deleteEvent(Event event){
        eventRepository.delete(event);
    }

    public String getPriority(Long event_id){

        return eventRepository.getPriority(userService.getCurrentUser().getId(),event_id);
    }

    public void changePriority(Long event_id, Long priority_id){

        eventRepository.changePriority(userService.getCurrentUser().getId(),event_id,priority_id);
    }

    public boolean isParticipant(Long event_id){

        return eventRepository.isParticipant(userService.getCurrentUser().getId(),event_id);
    }

    public void leaveEvent(Long event_id){

        eventRepository.deleteParticipant(userService.getCurrentUser().getId(),event_id);
    }

    public Long countPublic(String pattern, String category, LocalDateTime start, LocalDateTime finish) {
        return eventRepository.countPublic(pattern, start, finish, category);
    }

    public List<Event> searchPublic(String pattern, String category, LocalDateTime start, LocalDateTime finish,
                                                   Long limit, Long offset) {
        return eventRepository.searchPublic(pattern, start, finish, category, limit, offset);
    }

    public Long countUserEvents(String pattern, LocalDateTime start, LocalDateTime finish,
                                String category, Long userId, Long priority, Boolean byPriority,
                                Boolean privat) {
        return eventRepository.countUserEvents(pattern, start, finish, category,
                userId, priority, byPriority, privat);
    }

    public List<Event> searchUserEvents(String pattern, LocalDateTime start, LocalDateTime finish,
                                                             String category, Long userId, Long priority, Boolean byPriority,
                                                             Boolean privat, Long limit, Long offset) {
        return eventRepository.searchUserEvents(pattern, start, finish, category,
                userId, priority, byPriority, privat, limit, offset);
    }

    public Long countCreated(String pattern, LocalDateTime start, LocalDateTime finish,
                                String category, Long userId, Long priority, Boolean byPriority) {
        return eventRepository.countCreated(pattern, start, finish, category,
                userId, priority, byPriority);
    }

    public List<Event> searchCreated(String pattern, LocalDateTime start, LocalDateTime finish,
                                        String category, Long userId, Long priority, Boolean byPriority,
                                        Long limit, Long offset) {
        return eventRepository.searchCreated(pattern, start, finish, category,
                userId, priority, byPriority, limit, offset);
    }

    public Long countDrafts(String pattern, LocalDateTime start, LocalDateTime finish,
                             String category, Long userId, Long priority, Boolean byPriority) {
        return eventRepository.countDrafts(pattern, start, finish, category,
                userId, priority, byPriority);
    }

    public List<Event> searchDrafts(String pattern, LocalDateTime start, LocalDateTime finish,
                                     String category, Long userId, Long priority, Boolean byPriority,
                                     Long limit, Long offset) {
        return eventRepository.searchDrafts(pattern, start, finish, category,
                userId, priority, byPriority, limit, offset);
    }

    public List<Category> getCategories(){

       return eventRepository.getCategories();
    }

    public List<User> getFriendsNotParticipants(Long id){

        List<User> candidates = new ArrayList<>();

        for (User friends: userService.getFriendsByUserId(userService.getCurrentUser().getId())) {

            if (!getParticipants(id).contains(friends)) candidates.add(friends);
            System.out.println(getParticipants(id).contains(friends));
        }
        System.out.println(candidates);
        return candidates;
    }

    public Map<Integer, List<Integer>> getCalendarCounts(LocalDateTime start, LocalDateTime finish,
                                                         Long user_id, Boolean privat) {
        List<CalendarData> dataList = eventRepository.getCalendarData(start, finish, user_id, privat);
        Map<Integer, List<Integer>> result = new HashMap<>();
        for (LocalDateTime current = start; current.isBefore(finish); current = current.plusDays(1)) {
            LocalDateTime currentStart = current;
            LocalDateTime currentFinish = current.plusDays(1);
            List<CalendarData> dataByDay = dataList.stream()
                    .filter(event -> event.getStart().isBefore(currentFinish))
                    .filter(event -> event.getFinish().isAfter(currentStart))
                    .collect(Collectors.toList());
            result.put(current.getDayOfMonth(), Arrays.asList(
                    (int) dataByDay.stream().filter(event -> event.getPriority() == 0L).count(),
                    (int) dataByDay.stream().filter(event -> event.getPriority() == 1L).count(),
                    (int) dataByDay.stream().filter(event -> event.getPriority() == 2L).count(),
                    privat ? (int) dataByDay.stream().filter(event -> event.getPriority() == 3L).count() : 0
            ));
        }
        return result;
    }

    public void removeParticipants(List<User> users,Long id){

        for (User user:users) {
            eventRepository.deleteParticipant(user.getId(),id);
        }
    }
}

File: client/src/app/services/event.service.ts
import {Injectable} from '@angular/core';
import {HttpClient, HttpResponse} from '@angular/common/http';
import {Event} from '../model/event';
import {Observable} from "rxjs/Observable";
import {Router} from '@angular/router';
import {User} from "../model/user";
import {Category} from "../model/category";

@Injectable()
export class EventService {

  private base_url = '/event/';


  constructor(private http: HttpClient, private router: Router) {

  }

  getEventById(id) {
    return this.http.get(this.base_url + id);
  }

  public createEvent(event) {

   return this.http.post(this.base_url, event);

  }

  public getEvent(id: number): Observable<Event> {
    return this.http.get<Event>(this.base_url + id);
  }

  public updateEvent(event: Event): Observable<Object> {
    return this.http.post(this.base_url + event.id, event);
  }

  public getPublicEvents(): Observable<Event[]> {
    return this.http.get<Event[]>(this.base_url);
  }

  public getPriority(id) {

    return this.http.get(this.base_url + id + "/priority", {responseType: 'text'});
  }

  public changePriority(id,priority_id) {

    return this.http.get(this.base_url + id + "/priority/change",
      {
        params: {
          priority_id:priority_id
        }
      }
    )
  }
  public isParticipantRequest(id) {
    return this.http.get(this.base_url + id + "/isParticipant", {responseType: 'text'});
  }

  public joinToEvent(id){
    return this.http.get(this.base_url + id + "/join");
  }

  public getParticipants(id){
    return this.http.get(this.base_url+id+"/participants")
  }

  public leaveEvent(id){
    return this.http.get(this.base_url+id+"/leave")
  }

  public deleteEvent(id){
    return this.http.delete(this.base_url+id)
  }

  public addUsers(users:User[],id){
    return this.http.post(this.base_url+id+"/participants",users)
  }

  public getCategories(): Observable<Category[]>{
    return this.http.get<Category[]>(this.base_url+"categories")
  }

  public getCalendarData(year: number, month: number, id: number, privat: boolean) {
    console.log(privat);
    return this.http.get(this.base_url + `${id}/calendar/${year}/${month+1}?privat=${privat}`);
  }

  public getFilteredEvents(pattern: string, category: string, start: Date, finish: Date, limit: number, offset: number):Observable<HttpResponse<Event[]>> {
    return this.http.get<Event[]>(this.base_url + 'filter' +
      `?pattern=${pattern.toLowerCase()}&category=${category}&start=${start.toISOString()}&finish=${finish.toISOString()}&limit=${limit}&offset=${offset}`, {observe: 'response'});
  }

  public getFilteredUserEvents(pattern: string, category: string, start: Date, finish: Date,
      user_id: number, priority: number, by_priority: boolean, privat: boolean,
      limit: number, offset: number):Observable<HttpResponse<Event[]>> {
    console.log(privat);
    return this.http.get<Event[]>(this.base_url + `user/${user_id}/filter` +
      `?pattern=${pattern.toLowerCase()}&category=${category}&start=${start.toISOString()}&finish=${finish.toISOString()}` +
      `&priority=${priority}&byPriority=${by_priority}&privat=${privat}` +
      `&limit=${limit}&offset=${offset}`, {observe: 'response'});
  }
}
